### PR TITLE
Fix multi-threaded use of HttpUrlConnection

### LINKED ISSUE #467: HttpUrlConnection instrumentation in multi-threaded use cases
When starting a span in one therad via `HttpUrlConnection#connect()` and ending it in another thread via `HttpUrlConnection#getInputStream()`, the span's activation and deactivation happen in different threads, corrupting the activation state.

### PR DESCRIPTION
**Pull request by felixbarny**
fixes elastic/apm-agent-java#467

### CODE REVIEW

File: apm-agent-plugins/apm-urlconnection-plugin/src/main/java/co/elastic/apm/agent/urlconnection/HttpUrlConnectionInstrumentation.java [hunk 1]
```diff
@@ -63,43 +67,52 @@
...
                                 @Nullable @Advice.Thrown Throwable t,
-                                @Advice.FieldValue("responseCode") int responseCode) {
-            if (tracer == null) {
+                                @Advice.FieldValue("responseCode") int responseCode,
+                                @Nullable @Advice.Local("span") Span span,
+                                @Advice.Origin String signature) {
+            if (span == null) {
                 return;
+            } else {
             }
```
Comment (eyalkoren) [inline apm-agent-plugins/apm-urlconnection-plugin/src/main/java/co/elastic/apm/agent/urlconnection/HttpUrlConnectionInstrumentation.java]: ?

File: apm-agent-plugins/apm-urlconnection-plugin/src/main/java/co/elastic/apm/agent/urlconnection/HttpUrlConnectionInstrumentation.java [hunk 2]
```diff
@@ -121,15 +134,9 @@ public static void exit(@Advice.This HttpURLConnection thiz,
...
-            if (tracer == null) {
-                return;
-            }
-            final TraceContextHolder<?> active = tracer.getActive();
-            if (active instanceof Span) {
-                final Span span = (Span) active;
-                if (span.isOriginatedBy(thiz)) {
-                    span.captureException(t).deactivate().end();
-                }
+            Span span = inFlightSpans.get(thiz);
```
Comment (eyalkoren) [inline apm-agent-plugins/apm-urlconnection-plugin/src/main/java/co/elastic/apm/agent/urlconnection/HttpUrlConnectionInstrumentation.java]: Don't we want to remove from the map?

**Comment by eyalkoren** [comment]
Does it fix the Tomcat 7 IT?

File: apm-agent-plugins/apm-urlconnection-plugin/src/main/java/co/elastic/apm/agent/urlconnection/HttpUrlConnectionInstrumentation.java (hunk 2 above)
Comment (felixbarny) [reply apm-agent-plugins/apm-urlconnection-plugin/src/main/java/co/elastic/apm/agent/urlconnection/HttpUrlConnectionInstrumentation.java]: good catch!

File: apm-agent-plugins/apm-urlconnection-plugin/src/main/java/co/elastic/apm/agent/urlconnection/HttpUrlConnectionInstrumentation.java (hunk 1 above)
Comment (felixbarny) [reply apm-agent-plugins/apm-urlconnection-plugin/src/main/java/co/elastic/apm/agent/urlconnection/HttpUrlConnectionInstrumentation.java]: very good question 😄

**Comment by felixbarny** [comment]
> Does it fix the Tomcat 7 IT?
No, I don't think so. The Tomcat test failure was related to the Executor instrumentation.

**Pull request merged**